diff --git a/modules/svg/include/SkSVGCircle.h b/modules/svg/include/SkSVGCircle.h
index 8bc0da9311..f0fe478200 100644
--- a/modules/svg/include/SkSVGCircle.h
+++ b/modules/svg/include/SkSVGCircle.h
@@ -30,6 +30,8 @@ protected:
 
     SkPath onAsPath(const SkSVGRenderContext&) const override;
 
+    SkRect onObjectBoundingBox(const SkSVGRenderContext&) const override;
+
 private:
     SkSVGCircle();
 
diff --git a/modules/svg/include/SkSVGPath.h b/modules/svg/include/SkSVGPath.h
index 169837e3cc..d92447ca76 100644
--- a/modules/svg/include/SkSVGPath.h
+++ b/modules/svg/include/SkSVGPath.h
@@ -26,6 +26,8 @@ protected:
 
     SkPath onAsPath(const SkSVGRenderContext&) const override;
 
+    SkRect onObjectBoundingBox(const SkSVGRenderContext&) const override;
+
 private:
     SkSVGPath();
 
diff --git a/modules/svg/include/SkSVGPoly.h b/modules/svg/include/SkSVGPoly.h
index ad890aabee..d3a314134b 100644
--- a/modules/svg/include/SkSVGPoly.h
+++ b/modules/svg/include/SkSVGPoly.h
@@ -34,6 +34,8 @@ protected:
 
     SkPath onAsPath(const SkSVGRenderContext&) const override;
 
+    SkRect onObjectBoundingBox(const SkSVGRenderContext&) const override;
+
 private:
     SkSVGPoly(SkSVGTag);
 
diff --git a/modules/svg/src/SkSVGCircle.cpp b/modules/svg/src/SkSVGCircle.cpp
index 6b25f0980d..208fa3b504 100644
--- a/modules/svg/src/SkSVGCircle.cpp
+++ b/modules/svg/src/SkSVGCircle.cpp
@@ -74,3 +74,11 @@ SkPath SkSVGCircle::onAsPath(const SkSVGRenderContext& ctx) const {
 
     return path;
 }
+
+SkRect SkSVGCircle::onObjectBoundingBox(const SkSVGRenderContext& ctx) const {
+    auto& lctx = ctx.lengthContext();
+    const auto cx = lctx.resolve(fCx, SkSVGLengthContext::LengthType::kHorizontal);
+    const auto cy = lctx.resolve(fCy, SkSVGLengthContext::LengthType::kVertical);
+    const auto  r = lctx.resolve(fR , SkSVGLengthContext::LengthType::kOther);
+    return { cx - r, cy - r, cx + r, cy + r };
+}
diff --git a/modules/svg/src/SkSVGPath.cpp b/modules/svg/src/SkSVGPath.cpp
index f83937901d..ee8a7796cd 100644
--- a/modules/svg/src/SkSVGPath.cpp
+++ b/modules/svg/src/SkSVGPath.cpp
@@ -39,3 +39,7 @@ SkPath SkSVGPath::onAsPath(const SkSVGRenderContext& ctx) const {
     this->mapToParent(&path);
     return path;
 }
+
+SkRect SkSVGPath::onObjectBoundingBox(const SkSVGRenderContext& ctx) const {
+    return fPath.getBounds();
+}
diff --git a/modules/svg/src/SkSVGPoly.cpp b/modules/svg/src/SkSVGPoly.cpp
index 0aac9e6783..6079a86c3f 100644
--- a/modules/svg/src/SkSVGPoly.cpp
+++ b/modules/svg/src/SkSVGPoly.cpp
@@ -46,3 +46,7 @@ SkPath SkSVGPoly::onAsPath(const SkSVGRenderContext& ctx) const {
     this->mapToParent(&path);
     return path;
 }
+
+SkRect SkSVGPoly::onObjectBoundingBox(const SkSVGRenderContext& ctx) const {
+    return fPath.getBounds();
+}
diff --git a/modules/svg/src/SkSVGUse.cpp b/modules/svg/src/SkSVGUse.cpp
index bf4457c27a..f8def4c8b5 100644
--- a/modules/svg/src/SkSVGUse.cpp
+++ b/modules/svg/src/SkSVGUse.cpp
@@ -94,5 +94,5 @@ SkRect SkSVGUse::onObjectBoundingBox(const SkSVGRenderContext& ctx) const {
     const SkRect bounds = ref->objectBoundingBox(ctx);
     const SkScalar x = ctx.lengthContext().resolve(fX, SkSVGLengthContext::LengthType::kHorizontal);
     const SkScalar y = ctx.lengthContext().resolve(fY, SkSVGLengthContext::LengthType::kVertical);
-    return SkRect::MakeXYWH(x, y, bounds.width(), bounds.height());
+    return SkRect::MakeXYWH(x + bounds.left(), y + bounds.top(), bounds.width(), bounds.height());
 }
