diff --git a/modules/svg/src/SkSVGUse.cpp b/modules/svg/src/SkSVGUse.cpp
index bf4457c27a..f8def4c8b5 100644
--- a/modules/svg/src/SkSVGUse.cpp
+++ b/modules/svg/src/SkSVGUse.cpp
@@ -94,5 +94,5 @@ SkRect SkSVGUse::onObjectBoundingBox(const SkSVGRenderContext& ctx) const {
     const SkRect bounds = ref->objectBoundingBox(ctx);
     const SkScalar x = ctx.lengthContext().resolve(fX, SkSVGLengthContext::LengthType::kHorizontal);
     const SkScalar y = ctx.lengthContext().resolve(fY, SkSVGLengthContext::LengthType::kVertical);
-    return SkRect::MakeXYWH(x, y, bounds.width(), bounds.height());
+    return SkRect::MakeXYWH(x + bounds.left(), y + bounds.top(), bounds.width(), bounds.height());
 }
